README.md
# Software Component : Example

This repo uses the common data model, interfaces, build system, and component
launcher provided by the standalone **databus** project repo.

It's primary purpose to exercise and illustrate how to build  components (i.e. 
applications) for various RTI Connext SDKs using a single `CMakeLists.txt` file.

- RTI Connext DDS Professional (*pro*)
- RTI Connext DDS Micro (*micro*)

The example consists of the following applications

- `pro_publisher` : publishes dummy data to the `Camera` topic using *pro*
- `micro_subscriber` : subscribes to the `Camera` topic using *micro*

## Assumptions

- `$DATABUSHOME` refers to the location of the **databus** project repository

## Files Overview

The files in this repo are modified example files, generated using the `rtiddsgen` utility for the datatype `dds::sensing::Camera`, defined in `$DATABUSHOME/res/types/data/sensing/Camera_t.idl`.

### Applications

[pro_publisher.cxx](pro_pub/pro_publisher.cxx):

- the C++ publisher example generated by *RTI Connext DDS Professional*'s `rtiddsgen`
- the generated example has been modified to use the constants defined in `$DATABUSHOME/res/types/services/Drive_t.idl` for the `Camera` type and topic.

[micro_subscriber.c](micro_sub/micro_subscriber.c):

- the C subscriber example generated by *RTI Connext DDS Micro*'s `rtiddsgen`
- the generated example has been modified to use the constants defined in `$DATABUSHOME/res/types/services/Drive_t.idl` for the `Camera` type and topic.

# Building the Applications

This repo uses the *common build system* provided by the databus project repo located at `$DATABUSHOME`.

## Generate the build system for an RTI Connext SDK

From this git repo's top-level directory:

- To generate the build system for *RTI Connext DDS Professional*, e.g.

      $DATABUSHOME/bin/build-gen.sh pro x64Linux4gcc7.3.0 Debug

- To generate the build system for *RTI Connext DDS Micro*, e.g.

      $DATABUSHOME/bin/build-gen.sh micro x64Linux4gcc7.3.0 Release

This step creates a `build/` directory in the project top-level directory. The `build/` directory contains a shell script to build for the specified target platform and build kind.

For more details, please refer to the documentation on the *common build system* located at `$DATABUSHOME/doc/Build.md`.

## Building this repo using the generated build system

From this git repo's top-level directory, run the generated build script to build the datatypes:

- To build for *RTI Connext DDS Professional*, e.g.:

      ./build/pro-x64Linux4gcc7.3.0-Debug.sh

- To build for *RTI Connext DDS Micro*, e.g.:

      ./build/micro-x64Linux4gcc7.3.0-Release.sh



# Running the Applications

This repo uses the *common component launcher* provided by the databus project repo located at `$DATABUSHOME`. For documentation on the `$DATABUSHOME/bin/run` launcher utility, please refer to the documentation located at `$DATABUSHOME/doc/Run.md`.

Open two terminal windows...

- Launch the `pro_publisher` in terminal #1 

      $DATABUSHOME/bin/run Drive build/pro/x64Linux4gcc7.3.0/Debug/pro_publisher [domain_id] [sample_count]

   You should see output like:

        :
        Writing Camera, count 0
        Writing Camera, count 1
        Writing Camera, count 3
        :

- Launch the `micro_subscriber` in terminal #2

      $DATABUSHOME/bin/run Drive build/micro/x64Linux4gcc7.3.0/Release/micro_subscriber [-domain <id>]

   You should see output like:

        Running for 24 hours, press Ctrl-C to exit
        Matched a publisher

        Valid sample received

        Valid sample received
        
        :
        
Press ^C in a terminal window to terminate.

---
(C) Copyright 2020-2022 Real-Time Innovations, Inc.  All rights reserved.

The use of this software is governed by the terms specified in the RTI Labs License Agreement, available at https://www.rti.com/terms/RTILabs. 

By accessing, downloading, or otherwise using this software, you agree to be bound by those terms.


ReleaseNotes.md
# Release Notes

# v0.2

- Update for constant name changes in the upstream databus repo

# v0.1

- Build Connext Pro, Connext Micro, and Connext Micro 2 applications in a
  single project using CMake
- Rudimentary example with minimal modifications to the
  *rtiddsgen* generated code

---
(C) Copyright 2020-2022 Real-Time Innovations, Inc.  All rights reserved.

The use of this software is governed by the terms specified in the RTI Labs License Agreement, available at https://www.rti.com/terms/RTILabs. 

By accessing, downloading, or otherwise using this software, you agree to be bound by those terms.


micro_sub/Application.c
/******************************************************************************
(C) Copyright 2020-2022 Real-Time Innovations, Inc.  All rights reserved.

The use of this software is governed by the terms specified in the 
RTI Labs License Agreement, available at https://www.rti.com/terms/RTILabs. 

By accessing, downloading, or otherwise using this software, you agree to 
be bound by those terms.
*****************************************************************************/

#include "Application.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "rti_me_c.h"
#include "disc_dpde/disc_dpde_discovery_plugin.h"
#include "res/types/services/Drive_t.h"
#include "res/types/data/sensing/Camera_tPlugin.h"
#include "wh_sm/wh_sm_history.h"
#include "rh_sm/rh_sm_history.h"
#include "netio/netio_udp.h"

void
Application_help(char *appname)
{
    printf("%s [options]\n", appname);
    printf("options:\n");
    printf("-h                 - This text\n");
    printf("-domain <id>       - DomainId (default: 0)\n");
    printf("-udp_intf <intf>   - udp interface (no default)\n");
    printf("-peer <address>    - peer address (no default)\n");
    printf("-count <count>     - count (default -1)\n");
    printf("-sleep <ms>        - sleep between sends (default 1s)\n");
    printf("\n");
}

struct Application *
Application_create(const char *local_participant_name,
        const char *remote_participant_name,
        DDS_Long domain_id, char *udp_intf, char *peer,
        DDS_Long sleep_time, DDS_Long count)
{
    DDS_ReturnCode_t retcode;
    DDS_DomainParticipantFactory *factory = NULL;
    struct DDS_DomainParticipantFactoryQos dpf_qos =
        DDS_DomainParticipantFactoryQos_INITIALIZER;
    struct DDS_DomainParticipantQos dp_qos =
        DDS_DomainParticipantQos_INITIALIZER;
    DDS_Boolean success = DDS_BOOLEAN_FALSE;
    struct Application *application = NULL;
    RT_Registry_T *registry = NULL;
    struct UDP_InterfaceFactoryProperty *udp_property = NULL;

    struct DPDE_DiscoveryPluginProperty discovery_plugin_properties =
        DPDE_DiscoveryPluginProperty_INITIALIZER;

    /* Uncomment to increase verbosity level:
       OSAPI_Log_set_verbosity(OSAPI_LOG_VERBOSITY_WARNING);
       */
    application = (struct Application *)malloc(sizeof(struct Application));

    if (application == NULL)
    {
        printf("failed to allocate application\n");
        goto done;
    }

    application->sleep_time = sleep_time;
    application->count = count;

    factory = DDS_DomainParticipantFactory_get_instance();

    registry =
        DDS_DomainParticipantFactory_get_registry
        (DDS_DomainParticipantFactory_get_instance());

    if (!RT_Registry_register(registry, DDSHST_WRITER_DEFAULT_HISTORY_NAME,
                WHSM_HistoryFactory_get_interface(), NULL, NULL))
    {
        printf("failed to register wh\n");
        goto done;
    }

    if (!RT_Registry_register(registry, DDSHST_READER_DEFAULT_HISTORY_NAME,
                RHSM_HistoryFactory_get_interface(), NULL, NULL))
    {
        printf("failed to register rh\n");
        goto done;
    }

    /* Configure UDP transport's allowed interfaces */
    if (!RT_Registry_unregister(registry, NETIO_DEFAULT_UDP_NAME, NULL, NULL))
    {
        printf("failed to unregister udp\n");
        goto done;
    }

    udp_property = (struct UDP_InterfaceFactoryProperty *)
        malloc(sizeof(struct UDP_InterfaceFactoryProperty));
    if (udp_property == NULL)
    {
        printf("failed to allocate udp properties\n");
        goto done;
    }
    *udp_property = UDP_INTERFACE_FACTORY_PROPERTY_DEFAULT;

    /* For additional allowed interface(s), increase maximum and length, and
       set interface below:
       */
    if (!REDA_StringSeq_set_maximum(&udp_property->allow_interface,2))
    {
        printf("failed to set allow_interface maximum\n");
        goto done;
    }
    if (!REDA_StringSeq_set_length(&udp_property->allow_interface,2))
    {
        printf("failed to set allow_interface length\n");
        goto done;
    }

    /* loopback interface */
#if defined(RTI_DARWIN)
    *REDA_StringSeq_get_reference(&udp_property->allow_interface,0) = 
        DDS_String_dup("lo0");
#elif defined (RTI_LINUX)
    *REDA_StringSeq_get_reference(&udp_property->allow_interface,0) = 
        DDS_String_dup("lo");
#elif defined (RTI_VXWORKS)
    *REDA_StringSeq_get_reference(&udp_property->allow_interface,0) = 
        DDS_String_dup("lo0");
#elif defined(RTI_WIN32)
    *REDA_StringSeq_get_reference(&udp_property->allow_interface,0) = 
        DDS_String_dup("Loopback Pseudo-Interface 1");
#else
    *REDA_StringSeq_get_reference(&udp_property->allow_interface,0) = 
        DDS_String_dup("lo");
#endif

    if (udp_intf != NULL)
    { /* use interface supplied on command line */
        *REDA_StringSeq_get_reference(&udp_property->allow_interface,1) = 
            DDS_String_dup(udp_intf);
    } 
    else                /* use hardcoded interface */
    {
#if defined(RTI_DARWIN)
        *REDA_StringSeq_get_reference(&udp_property->allow_interface,1) = 
            DDS_String_dup("en1");
#elif defined (RTI_LINUX)
        *REDA_StringSeq_get_reference(&udp_property->allow_interface,1) = 
            DDS_String_dup("eth0");
#elif defined (RTI_VXWORKS)
        *REDA_StringSeq_get_reference(&udp_property->allow_interface,1) = 
            DDS_String_dup("geisc0");
#elif defined(RTI_WIN32)
        *REDA_StringSeq_get_reference(&udp_property->allow_interface,1) = 
            DDS_String_dup("Local Area Connection");
#else
        *REDA_StringSeq_get_reference(&udp_property->allow_interface,1) = 
            DDS_String_dup("ce0");
#endif
    }

    if (!RT_Registry_register(registry, NETIO_DEFAULT_UDP_NAME,
                UDP_InterfaceFactory_get_interface(),
                (struct RT_ComponentFactoryProperty*)udp_property, NULL))
    {
        printf("failed to register udp\n");
        goto done;
    }

    DDS_DomainParticipantFactory_get_qos(factory, &dpf_qos);
    dpf_qos.entity_factory.autoenable_created_entities = DDS_BOOLEAN_FALSE;
    DDS_DomainParticipantFactory_set_qos(factory, &dpf_qos);

    if (peer == NULL)
    {
        peer = "127.0.0.1"; /* default to loopback */
    }

    if (!RT_Registry_register(registry,
                "dpde",
                DPDE_DiscoveryFactory_get_interface(),
                &discovery_plugin_properties._parent, 
                NULL))
    {
        printf("failed to register dpde\n");
        goto done;
    }

    if (!RT_ComponentFactoryId_set_name(&dp_qos.discovery.discovery.name,"dpde"))
    {
        printf("failed to set discovery plugin name\n");
        goto done;
    }

    if (!DDS_StringSeq_set_maximum(&dp_qos.discovery.initial_peers,1))
    {
        printf("failed to set initial peers maximum\n");
        goto done;
    }
    if (!DDS_StringSeq_set_length(&dp_qos.discovery.initial_peers,1))
    {
        printf("failed to set initial peers length\n");
        goto done;
    }
    *DDS_StringSeq_get_reference(&dp_qos.discovery.initial_peers,0) = DDS_String_dup(peer);

    /* if there are more remote or local endpoints, you need to increase these limits */
    dp_qos.resource_limits.max_destination_ports = 32;
    dp_qos.resource_limits.max_receive_ports = 32;
    dp_qos.resource_limits.local_topic_allocation = 1;
    dp_qos.resource_limits.local_type_allocation = 1;
    dp_qos.resource_limits.local_reader_allocation = 1;
    dp_qos.resource_limits.local_writer_allocation = 1;
    dp_qos.resource_limits.remote_participant_allocation = 8;
    dp_qos.resource_limits.remote_reader_allocation = 8;
    dp_qos.resource_limits.remote_writer_allocation = 8;

    strcpy(dp_qos.participant_name.name, local_participant_name);
    
    application->participant =
        DDS_DomainParticipantFactory_create_participant(factory, domain_id,
                &dp_qos, NULL,
                DDS_STATUS_MASK_NONE);

    if (application->participant == NULL)
    {
        printf("failed to create participant\n");
        goto done;
    }

    sprintf(application->type_name, dds_Drive_Bus_main_registered_type_Camera);
    retcode = DDS_DomainParticipant_register_type(application->participant,
            application->type_name,
            dds_sensing_CameraTypePlugin_get());
    if (retcode != DDS_RETCODE_OK)
    {
        printf("failed to register type: %s\n", "test_type");
        goto done;
    }

    sprintf(application->topic_name, dds_Drive_Bus_main_topic_Camera );
    application->topic =
        DDS_DomainParticipant_create_topic(application->participant,
                application->topic_name,
                application->type_name,
                &DDS_TOPIC_QOS_DEFAULT, NULL,
                DDS_STATUS_MASK_NONE);

    if (application->topic == NULL)
    {
        printf("topic == NULL\n");
        goto done;
    }

    success = DDS_BOOLEAN_TRUE;

done:
    DDS_DomainParticipantQos_finalize(&dp_qos);

    if (!success)
    {
        if (udp_property != NULL)
        {
            UDP_InterfaceFactoryProperty_finalize(udp_property);
            free(udp_property);
        }
        free(application);
        application = NULL;
    }

    return application;
}

DDS_ReturnCode_t
Application_enable(struct Application * application)
{
    DDS_Entity *entity;
    DDS_ReturnCode_t retcode;

    entity = DDS_DomainParticipant_as_entity(application->participant);

    retcode = DDS_Entity_enable(entity);
    if (retcode != DDS_RETCODE_OK)
    {
        printf("failed to enable entity\n");
    }

    return retcode;
}

void
Application_delete(struct Application *application)
{
    DDS_ReturnCode_t retcode;
    RT_Registry_T *registry = NULL;
    struct UDP_InterfaceFactoryProperty *udp_property = NULL;

    retcode = DDS_DomainParticipant_delete_contained_entities(application->participant);
    if (retcode != DDS_RETCODE_OK)
    {
        printf("failed to delete contained entities (retcode=%d)\n",retcode);
    }

    retcode =
        DDS_DomainParticipantFactory_delete_participant
        (DDS_DomainParticipantFactory_get_instance(), application->participant);
    if (retcode != DDS_RETCODE_OK)
    {
        printf("failed to delete participant: %d\n", retcode);
        return;
    }

    registry = DDS_DomainParticipantFactory_get_registry
        (DDS_DomainParticipantFactory_get_instance());

    if (!RT_Registry_unregister(registry, NETIO_DEFAULT_UDP_NAME,
                (struct RT_ComponentFactoryProperty**)&udp_property, 
                NULL))
    {
        printf("failed to unregister udp\n");
        return;
    }
    if (udp_property != NULL)
    {
        UDP_InterfaceFactoryProperty_finalize(udp_property);
        free(udp_property);
        udp_property = NULL;
    }
    if (!RT_Registry_unregister(registry, "dpde", NULL, NULL))
    {
        printf("failed to unregister dpde\n");
        return;
    }
    if (!RT_Registry_unregister(registry, DDSHST_READER_DEFAULT_HISTORY_NAME, NULL, NULL))
    {
        printf("failed to unregister rh\n");
        return;
    }
    if (!RT_Registry_unregister(registry, DDSHST_WRITER_DEFAULT_HISTORY_NAME, NULL, NULL))
    {
        printf("failed to unregister wh\n");
        return;
    }

    free(application);

    retcode = DDS_DomainParticipantFactory_finalize_instance();
    if (retcode != DDS_RETCODE_OK)
    {
        printf("failed to finalize instance %d\n", retcode);
        return;
    }
}


micro_sub/Application.h
/******************************************************************************
(C) Copyright 2020-2022 Real-Time Innovations, Inc.  All rights reserved.

The use of this software is governed by the terms specified in the 
RTI Labs License Agreement, available at https://www.rti.com/terms/RTILabs. 

By accessing, downloading, or otherwise using this software, you agree to 
be bound by those terms.
*****************************************************************************/

#ifndef Application_h
#define Application_h

#include "rti_me_c.h"

/*e \dref_Example_Config_UseReliableQos
* Define USE_RELIABLE_QOS to use reliable
* reliability on the DataReader; otherwise, use
* best-effort reliability by default.
*/

#define USE_RELIABLE_QOS

/*e \dref_Example_Config_UseSampleFilter
* Define USE_SAMPLE_FILTER to filter samples
* using call-backs on the DataReader's listener
* and, in case the first field of your type is not
* a DDS_Long called "id", modify
* "CameraImageSubscriber_filter_sample",
* "CameraImageSubscriber_on_before_sample_deserialize",
* "CameraImageSubscriber_deserialize_unsigned_long", and
* "CameraImageSubscriber_filter_sample" functions to
* correctly filter by the field you want to use for filtering.
*/

/*#define USE_SAMPLE_FILTER*/

/*e \dref_Example_Config_FilterOnDeserialize
* Define FILTER_ON_DESERIALIZE to enable
* filtering on call-back on_before_sample_deserialize;
* otherwise use call-back on_before_sample_commit
* by default. */

/*#define FILTER_ON_DESERIALIZE*/

struct Application {
    DDS_DomainParticipant *participant;
    char topic_name[255];
    char type_name[255];
    DDS_Long sleep_time;
    DDS_Long count;
    DDS_Topic *topic;
};

extern void 
Application_help(char *appname);

extern struct Application*
Application_create(const char *local_participant_name,
        const char *remote_participant_name,
        DDS_Long domain_id, char *udp_intf, char *peer, 
        DDS_Long sleep_time, DDS_Long count);

extern DDS_ReturnCode_t
Application_enable(struct Application *application);

extern void
Application_delete(struct Application *application);

#endif


micro_sub/micro_subscriber.c
/******************************************************************************
(C) Copyright 2020-2022 Real-Time Innovations, Inc.  All rights reserved.

The use of this software is governed by the terms specified in the 
RTI Labs License Agreement, available at https://www.rti.com/terms/RTILabs. 

By accessing, downloading, or otherwise using this software, you agree to 
be bound by those terms.
*****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "Application.h"
#include "rti_me_c.h"
#include "wh_sm/wh_sm_history.h"
#include "rh_sm/rh_sm_history.h"

#include "res/types/data/sensing/Camera_t.h"
#include "res/types/data/sensing/Camera_tSupport.h"
#include "res/types/data/sensing/Camera_tPlugin.h"

/* See Wire Protocol Specification on http://www.omg.org/spec/DDSI-RTPS/2.2/
 * for more details about CDR encapsulation.
 */

/*ci \brief Unsigned Long type size */
#define UNSIGNED_LONG_SIZE  4

void
dds_sensing_CameraSubscriber_on_data_available(void *listener_data,
        DDS_DataReader * reader)
{
    dds_sensing_CameraDataReader *hw_reader = dds_sensing_CameraDataReader_narrow(reader);
    DDS_ReturnCode_t retcode;
    struct DDS_SampleInfo *sample_info = NULL;
    dds_sensing_Camera *sample = NULL;

    struct DDS_SampleInfoSeq info_seq = 
        DDS_SEQUENCE_INITIALIZER;
    struct dds_sensing_CameraSeq sample_seq =
        DDS_SEQUENCE_INITIALIZER;

    const DDS_Long TAKE_MAX_SAMPLES = 32;
    DDS_Long i;

    retcode = dds_sensing_CameraDataReader_take(hw_reader,
            &sample_seq, &info_seq, TAKE_MAX_SAMPLES, 
            DDS_ANY_SAMPLE_STATE, DDS_ANY_VIEW_STATE, DDS_ANY_INSTANCE_STATE);

    if (retcode != DDS_RETCODE_OK)
    {
        printf("failed to take data, retcode(%d)\n", retcode);
        goto done;
    }

    /* Print each valid sample taken */
    for (i = 0; i < dds_sensing_CameraSeq_get_length(&sample_seq); ++i)
    {
        sample_info = DDS_SampleInfoSeq_get_reference(&info_seq, i);

        if (sample_info->valid_data)
        {

            sample = dds_sensing_CameraSeq_get_reference(&sample_seq, i);
            printf("\nValid sample received\n");
            /* TODO read and process sample attributes here */ 

        }
        else
        {
            printf("\nSample received\n\tINVALID DATA\n");
        }
    }

    dds_sensing_CameraDataReader_return_loan(hw_reader, &sample_seq, &info_seq);

done:
    dds_sensing_CameraSeq_finalize(&sample_seq);
    DDS_SampleInfoSeq_finalize(&info_seq);
}

void
dds_sensing_CameraSubscriber_on_subscription_matched(void *listener_data,
        DDS_DataReader * reader,
        const struct
        DDS_SubscriptionMatchedStatus
        *status)
{
    if (status->current_count_change > 0)
    {
        printf("Matched a publisher\n");
    }
    else if (status->current_count_change < 0)
    {
        printf("Unmatched a publisher\n");
    }
}

#ifdef USE_SAMPLE_FILTER
#ifdef FILTER_ON_DESERIALIZE
/*i
 * \brief Helper function to deserialize an unsigned long
 *
 * \param[in]  src_buffer      Pointer to CDR stream buffer
 * \param[in]  need_byte_swap  Indicates whether it is needed to swap bytes
 * \param[out] instance        Deserialized unsigned long
 */
void
dds_sensing_CameraSubscriber_deserialize_unsigned_long(char **src_buffer,
        RTI_BOOL need_byte_swap,
        DDS_UnsignedLong *instance)
{
    RTI_INT32 i;
    if (need_byte_swap)
    {
        for (i = 3; i >= 0; --i)
        {
            *((RTI_INT8*)instance + i) = *((*src_buffer)++);
        }
    }
    else
    {
        *instance = *(RTI_UINT32*)(*src_buffer);
        (*src_buffer) += CDR_LONG_SIZE;
    }
}

/*i
 * \brief Implementation of \ref DDS_DataReaderListener::on_before_sample_deserialize
 */
DDS_Boolean
dds_sensing_CameraSubscriber_on_before_sample_deserialize(
        void *listener_data,
        DDS_DataReader *reader,
        struct NDDS_Type_Plugin *plugin,
        struct CDR_Stream_t *stream,
        DDS_Boolean *dropped)
{
    DDS_Boolean result = DDS_BOOLEAN_FALSE;
    DDS_Long id = 0;
    RTI_BOOL need_byte_swap = DDS_BOOLEAN_FALSE;
    char *src_buffer = NULL;

    need_byte_swap = stream->need_byte_swap;
    src_buffer = CDR_Stream_get_current_position_ptr(stream);

    /* deserialize only field 'id', instead of the whole sample */

    if (!CDR_Stream_check_size(stream, UNSIGNED_LONG_SIZE))
    {
        printf("Failed to deserialize id. The stream is too short, missing data\n");
        goto done;
    }

    /* Note primitive types must be aligned to their length in the CDR stream.
     * For example, a long must start on a 4-byte boundary. The boundaries are
     * counted from the start of the CDR stream.
     * As the sample 'id' is the first data in the stream it is already aligned.
     * Position 0 (beginning of the stream) is aligned to 4 (size of long).
     *
     * NOTE: If you want to use a different field for filtering (e.g. you type does
     * not have a field called id as first field), you will need to reimplement this
     * function and dds_sensing_CameraSubscriber_deserialize_unsigned_long
     * to match your type.
     */
    dds_sensing_CameraSubscriber_deserialize_unsigned_long(&src_buffer,
            need_byte_swap,
            (DDS_UnsignedLong*)&id);

    /* filter the sample */
    *dropped = (id % 2 == 0) ? DDS_BOOLEAN_TRUE : DDS_BOOLEAN_FALSE;

    if (*dropped)
    {
        printf("\nSample filtered, before deserialize...\n\tDROPPED - id: %d\n",
                id);
    }

    result = DDS_BOOLEAN_TRUE;
done:
    return result;
}

#else

/*i
 * \brief Helper function to filter an dds_sensing_Camera sample
 *
 * \param[in]  sample       A dds_sensing_Camera data sample to filter
 * \param[out] drop_sample  Out parameter determining whether the sample
 *                          should be filtered out or not.
 */
void 
dds_sensing_CameraSubscriber_filter_sample(dds_sensing_Camera *sample,
        DDS_Boolean *drop_sample)
{
    /* Example filter: drop samples with even-numbered count in id */
    /* NOTE: If field "id" does not exist in your data type, change "id"
     * by the field you want to use for filtering.
     */
    *drop_sample = (sample->id % 2 == 0) ? DDS_BOOLEAN_TRUE : DDS_BOOLEAN_FALSE;
}

/*i
 * \brief Implementation of \ref DDS_DataReaderListener::on_before_sample_commit
 */
DDS_Boolean
dds_sensing_CameraSubscriber_on_before_sample_commit(
        void *listener_data,
        DDS_DataReader *reader,
        const void *const sample,
        const struct DDS_SampleInfo *const sample_info,
        DDS_Boolean *dropped)
{
    dds_sensing_Camera *hw_sample = (dds_sensing_Camera *)sample;

    dds_sensing_CameraSubscriber_filter_sample(hw_sample, dropped);

    if (*dropped)
    {
        printf("\nSample filtered, before commit...\n");
    }

    return DDS_BOOLEAN_TRUE;
}
#endif /* FILTER_ON_DESERIALIZE */
#endif /* USE_SAMPLE_FILTER */

int
subscriber_main_w_args(DDS_Long domain_id, char *udp_intf, char *peer,
        DDS_Long sleep_time, DDS_Long count)
{
    DDS_Subscriber *subscriber;
    DDS_DataReader *datareader;
    struct DDS_DataReaderQos dr_qos = DDS_DataReaderQos_INITIALIZER;
    DDS_ReturnCode_t retcode;
    struct DDS_DataReaderListener dr_listener =
        DDS_DataReaderListener_INITIALIZER;
    struct Application *application;

    application = Application_create(
            "Example (micro)", 
            "Example",
            domain_id, udp_intf, peer, sleep_time, count);

    if (application == NULL)
    {
        return 0;
    }

    subscriber =
        DDS_DomainParticipant_create_subscriber(application->participant,
                &DDS_SUBSCRIBER_QOS_DEFAULT,
                NULL, DDS_STATUS_MASK_NONE);
    if (subscriber == NULL)
    {
        printf("subscriber == NULL\n");
        goto done;
    }

#ifdef USE_SAMPLE_FILTER
    /* choose one callback to enable */
#ifdef FILTER_ON_DESERIALIZE
    dr_listener.on_before_sample_deserialize =
        dds_sensing_CameraSubscriber_on_before_sample_deserialize;
#else
    dr_listener.on_before_sample_commit = 
        dds_sensing_CameraSubscriber_on_before_sample_commit;
#endif

#endif

    dr_listener.on_data_available = dds_sensing_CameraSubscriber_on_data_available;
    dr_listener.on_subscription_matched =
        dds_sensing_CameraSubscriber_on_subscription_matched;

    /* Publisher sends samples with id = 0 or id = 1, so 2 instances maximum.
     * But in case filtering is done, all samples with 'id = 0' are
     * filtered so only one instance is needed.
     */
#ifdef USE_SAMPLE_FILTER
    dr_qos.resource_limits.max_instances = 1;
#else
    dr_qos.resource_limits.max_instances = 2;
#endif

    dr_qos.resource_limits.max_samples_per_instance = 32;
    dr_qos.resource_limits.max_samples = dr_qos.resource_limits.max_instances *
        dr_qos.resource_limits.max_samples_per_instance;
    /* if there are more remote writers, you need to increase these limits */
    dr_qos.reader_resource_limits.max_remote_writers = 10;
    dr_qos.reader_resource_limits.max_remote_writers_per_instance = 10;
    dr_qos.history.depth = 32;

    /* Reliability QoS */
#ifdef USE_RELIABLE_QOS
    dr_qos.reliability.kind = DDS_RELIABLE_RELIABILITY_QOS;
#else
    dr_qos.reliability.kind = DDS_BEST_EFFORT_RELIABILITY_QOS;
#endif

    datareader = DDS_Subscriber_create_datareader(subscriber,
            DDS_Topic_as_topicdescription
            (application->topic), &dr_qos,
            &dr_listener,
            DDS_DATA_AVAILABLE_STATUS |
            DDS_SUBSCRIPTION_MATCHED_STATUS);

    if (datareader == NULL)
    {
        printf("datareader == NULL\n");
        goto done;
    }

    retcode = Application_enable(application);
    if (retcode != DDS_RETCODE_OK)
    {
        printf("failed to enable application\n");
        goto done;
    }

    if (application->count != 0)
    {
        printf("Running for %d seconds, press Ctrl-C to exit\n",
                application->count);
        OSAPI_Thread_sleep(application->count * 1000);
    }
    else
    {
        int sleep_loop_count =  (24 * 60 * 60) / 2000;
        int sleep_loop_left = (24 * 60 * 60) % 2000;

        printf("Running for 24 hours, press Ctrl-C to exit\n");

        while (sleep_loop_count)
        {
            OSAPI_Thread_sleep(2000  * 1000); /* sleep is in ms */
            --sleep_loop_count;
        }

        OSAPI_Thread_sleep(sleep_loop_left * 1000);
    }

done:

    Application_delete(application);

    DDS_DataReaderQos_finalize(&dr_qos);

    return 0;
}

#if !(defined(RTI_VXWORKS) && !defined(__RTP__))
int
main(int argc, char **argv)
{
    DDS_Long i = 0;
    DDS_Long domain_id = 0;
    char *peer = NULL;
    char *udp_intf = NULL;
    DDS_Long sleep_time = 1000;
    DDS_Long count = 0;

    for (i = 1; i < argc; ++i)
    {
        if (!strcmp(argv[i], "-domain"))
        {
            ++i;
            if (i == argc)
            {
                printf("-domain <domain_id>\n");
                return -1;
            }
            domain_id = strtol(argv[i], NULL, 0);
        }
        else if (!strcmp(argv[i], "-udp_intf"))
        {
            ++i;
            if (i == argc)
            {
                printf("-udp_intf <interface>\n");
                return -1;
            }
            udp_intf = argv[i];
        }
        else if (!strcmp(argv[i], "-peer"))
        {
            ++i;
            if (i == argc)
            {
                printf("-peer <address>\n");
                return -1;
            }
            peer = argv[i];
        }
        else if (!strcmp(argv[i], "-sleep"))
        {
            ++i;
            if (i == argc)
            {
                printf("-sleep_time <sleep_time>\n");
                return -1;
            }
            sleep_time = strtol(argv[i], NULL, 0);
        }
        else if (!strcmp(argv[i], "-count"))
        {
            ++i;
            if (i == argc)
            {
                printf("-count <count>\n");
                return -1;
            }
            count = strtol(argv[i], NULL, 0);
        }
        else if (!strcmp(argv[i], "-h"))
        {
            Application_help(argv[0]);
            return 0;
        }
        else
        {
            printf("unknown option: %s\n", argv[i]);
            return -1;
        }
    }

    return subscriber_main_w_args(domain_id, udp_intf, peer, sleep_time, count);
}
#elif defined(RTI_VXWORKS)
int
subscriber_main(void)
{
    /* Explicitly configure args below */
    DDS_Long domain_id = 44;
    char *peer = "10.10.65.103";
    char *udp_intf = NULL;
    DDS_Long sleep_time = 1000;
    DDS_Long count = 0;

    return subscriber_main_w_args(domain_id, udp_intf, peer, sleep_time, count);
}
#endif
